pkg/timelock/timelock.go

```diff
@@ -273,26 +276,47 @@ func (tw *Worker) retrieveHistoricalLogs(ctx context.Context) (<-chan struct{},
 // processLogs is implemented as a fan-in for all the logs channels, merging all the data and handling logs sequentially.
 // This function is thread safe.
 func (tw *Worker) processLogs(ctx context.Context, oldLog, newLog <-chan types.Log) <-chan struct{} {
-	done := make(chan struct{})
+	var (
+		done, newDone, oldDone = make(chan struct{}), make(chan struct{}), make(chan struct{})
+		ctxwc, cancel          = context.WithCancel(ctx)
+	)
+
+	// Cancel the context and shutdown the processing routine if no more logs are available.
+	go func() {
+		defer cancel()
+		<-isclosed.All(ctxwc, oldDone, newDone)
+	}()
 
 	// This is the goroutine watching over the subscribed and historical logs.
 	go func() {
 		defer close(done)
 
 		for {
 			select {
-			case log := <-newLog:
-				if err := tw.handleLog(ctx, log); err != nil {
+			case log, open := <-newLog:
+				if !open {
+					close(newDone)
```

Comment:
if newLog is closed, signal new logs are done and set the channel to nil so this branch never is selected again